diff --git a/src/lib/libdwarf/dwarf_error.c b/src/lib/libdwarf/dwarf_error.c
index c961de6e..aab14aea 100644
--- a/src/lib/libdwarf/dwarf_error.c
+++ b/src/lib/libdwarf/dwarf_error.c
@@ -188,12 +188,12 @@ _dwarf_error_string(Dwarf_Debug dbg, Dwarf_Error * error,
         dbg->de_errhand(errptr, dbg->de_errarg);
         return;
     }
-    fflush(stderr);
-    fprintf(stderr,
-        "\nlibdwarf is unable to record error %s "
-        "No error argument or handler available\n",
-        dwarf_errmsg_by_number(errval));
-    fflush(stderr);
+    // fflush(stderr);
+    // fprintf(stderr,
+    //     "\nlibdwarf is unable to record error %s "
+    //     "No error argument or handler available\n",
+    //     dwarf_errmsg_by_number(errval));
+    // fflush(stderr);
     return;
 }
 
diff --git a/src/lib/libdwarf/dwarf_fill_in_attr_form.c b/src/lib/libdwarf/dwarf_fill_in_attr_form.c
index 98a16367..52b63516 100644
--- a/src/lib/libdwarf/dwarf_fill_in_attr_form.c
+++ b/src/lib/libdwarf/dwarf_fill_in_attr_form.c
@@ -90,7 +90,7 @@ _dwarf_fill_in_attr_form_abtable(Dwarf_CU_Context context,
             return DW_DLV_ERROR;
         }
     }
-    if (!abbrev_list->abl_attr || !abbrev_list->abl_form ) {
+    if (abbrev_list->abl_abbrev_count > 0 && (!abbrev_list->abl_attr || !abbrev_list->abl_form)) {
         dwarfstring m;
 
         dwarfstring_constructor(&m);
